presto-main/src/main/java/com/facebook/presto/operator/project/PageProcessor.java

```diff
@@ -76,42 +85,95 @@ public PageProcessor(Optional<PageFilter> filter, List<? extends PageProjection>
+    {
+        return pages.flatTransform(new PageFilterAndProject(session, yieldSignal, memoryContext));
+    }
+
+    private class SingletonPage
+            implements WorkProcessor.Process<Page>
+    {
+        private final LocalMemoryContext memoryContext;
+        private Page page;
 
-        if (page.getPositionCount() == 0) {
-            return EMPTY_PAGE_PROCESSOR_OUTPUT;
+        private SingletonPage(AggregatedMemoryContext memoryContext, Page page)
+        {
+            this.memoryContext = memoryContext.newLocalMemoryContext();
+            this.page = page;
+            this.memoryContext.setBytes(calculateRetainedSizeWithoutLoading(page));
         }
 
-        if (filter.isPresent()) {
-            SelectedPositions selectedPositions = filter.get().filter(session, filter.get().getInputChannels().getInputChannels(page));
-            if (selectedPositions.isEmpty()) {
-                return EMPTY_PAGE_PROCESSOR_OUTPUT;
+        @Override
+        public ProcessorState<Page> process()
+        {
+            if (page == null) {
+                return finished();
             }
 
-            if (projections.isEmpty()) {
-                return new PageProcessorOutput(() -> calculateRetainedSizeWithoutLoading(page), singletonIterator(Optional.of(new Page(selectedPositions.size()))));
+            Page result = page;
+            page = null;
+            memoryContext.close();
+            return ProcessorState.ofResult(result);
+        }
+    }
+
+    private class PageFilterAndProject
+            implements WorkProcessor.Transformation<Page, WorkProcessor<Page>>
+    {
+        private final ConnectorSession session;
+        private final DriverYieldSignal yieldSignal;
+        private final AggregatedMemoryContext memoryContext;
+
+        private PageFilterAndProject(ConnectorSession session, DriverYieldSignal yieldSignal, AggregatedMemoryContext memoryContext)
+        {
+            this.session = session;
+            this.yieldSignal = yieldSignal;
+            this.memoryContext = memoryContext;
+        }
+
+        @Override
+        public ProcessorState<WorkProcessor<Page>> process(Optional<Page> pageOptional)
+        {
+            if (!pageOptional.isPresent()) {
+                return finished();
             }
 
-            if (selectedPositions.size() != page.getPositionCount()) {
-                PositionsPageProcessorIterator pages = new PositionsPageProcessorIterator(session, yieldSignal, page, selectedPositions);
-                return new PageProcessorOutput(pages::getRetainedSizeInBytes, pages);
+            Page page = pageOptional.get();
+
+            // limit the scope of the dictionary ids to just one page
+            dictionarySourceIdFunction.reset();
+
+            if (page.getPositionCount() == 0) {
+                return needsMoreData();
```

Comment:
previously code returned `EMPTY_PAGE_PROCESSOR_OUTPUT` but not code is going to loop until it finds non empty page.